File: test/e2e/step_definitions/connect_steps.py
"""
Connect-Go protocol step definitions for E2E tests.
"""

import json
import requests
import concurrent.futures
from pytest_bdd import when, then, parsers


@when("I make a health check request using Connect-Go client")
def make_connect_health_request(server_url, http_client, context):
    """Make a health check request using Connect-Go protocol."""
    # Since we're using Python, we'll simulate the Connect-Go request
    # by making an HTTP POST request to the Connect endpoint
    url = f"{server_url}/health.v1.HealthService/Check"
    headers = {"Content-Type": "application/json"}
    data = "{}"

    try:
        response = http_client.post(url, data=data, headers=headers)
        context["response"] = response
        context["error"] = None
    except requests.exceptions.RequestException as e:
        context["response"] = None
        context["error"] = e


@when("I make a health check request using HTTP/JSON")
def make_http_health_request(server_url, http_client, context):
    """Make a health check request using HTTP/JSON."""
    url = f"{server_url}/health.v1.HealthService/Check"
    headers = {"Content-Type": "application/json"}
    data = "{}"

    try:
        response = http_client.post(url, data=data, headers=headers)
        context["response"] = response
        context["error"] = None
    except requests.exceptions.RequestException as e:
        context["response"] = None
        context["error"] = e


@when("I make a health check request with invalid content type")
def make_invalid_content_type_request(server_url, http_client, context):
    """Make a health check request with invalid content type."""
    url = f"{server_url}/health.v1.HealthService/Check"
    headers = {"Content-Type": "text/plain"}
    data = "{}"

    try:
        response = http_client.post(url, data=data, headers=headers)
        context["response"] = response
        context["error"] = None
    except requests.exceptions.RequestException as e:
        context["response"] = None
        context["error"] = e


@when("I make a request to a non-existent endpoint")
def make_nonexistent_endpoint_request(server_url, http_client, context):
    """Make a request to a non-existent endpoint."""
    url = f"{server_url}/health.v1.HealthService/NonExistentMethod"
    headers = {"Content-Type": "application/json"}
    data = "{}"

    try:
        response = http_client.post(url, data=data, headers=headers)
        context["response"] = response
        context["error"] = None
    except requests.exceptions.RequestException as e:
        context["response"] = None
        context["error"] = e


@when(parsers.parse("I make {num_requests:d} concurrent health check requests"))
def make_concurrent_requests(num_requests, server_url, context):
    """Make multiple concurrent health check requests."""
    url = f"{server_url}/health.v1.HealthService/Check"
    headers = {"Content-Type": "application/json"}
    data = "{}"

    def make_request():
        try:
            response = requests.post(url, data=data, headers=headers, timeout=5)
            return response, None
        except requests.exceptions.RequestException as e:
            return None, e

    # Make concurrent requests
    with concurrent.futures.ThreadPoolExecutor(max_workers=num_requests) as executor:
        futures = [executor.submit(make_request) for _ in range(num_requests)]
        results = [
            future.result() for future in concurrent.futures.as_completed(futures)
        ]

    context["concurrent_results"] = results


@then("the response should indicate SERVING status")
def check_serving_status(context):
    """Check that the response indicates SERVING status."""
    response = context.get("response")
    assert response is not None, "No response received"
    assert (
        response.status_code == 200
    ), f"Expected status 200, got {response.status_code}"

    try:
        json_data = response.json()
    except json.JSONDecodeError:
        assert False, "Response is not valid JSON"

    # Check for Connect-Go style response
    assert "status" in json_data, "Response does not contain 'status' field"
    assert (
        json_data["status"] == "SERVING_STATUS_SERVING"
    ), f"Expected SERVING_STATUS_SERVING, got {json_data['status']}"


@then("the response should not be empty")
def check_response_not_empty(context):
    """Check that the response is not empty."""
    response = context.get("response")
    assert response is not None, "No response received"
    assert response.content, "Response body is empty"


@then("the JSON response should contain SERVING_STATUS_SERVING")
def check_json_serving_status(context):
    """Check that the JSON response contains SERVING_STATUS_SERVING."""
    response = context.get("response")
    assert response is not None, "No response received"

    try:
        json_data = response.json()
    except json.JSONDecodeError:
        assert False, "Response is not valid JSON"

    assert "status" in json_data, "Response does not contain 'status' field"
    assert (
        json_data["status"] == "SERVING_STATUS_SERVING"
    ), f"Expected SERVING_STATUS_SERVING, got {json_data['status']}"


@then("the server should reject the request")
def check_request_rejected(context):
    """Check that the server rejected the request."""
    response = context.get("response")
    assert response is not None, "No response received"
    # For invalid content type, we expect 415 status code
    assert (
        response.status_code == 415
    ), f"Expected status 415 for rejected request, got {response.status_code}"


@then("all requests should succeed")
def check_all_requests_succeed(context):
    """Check that all concurrent requests succeeded."""
    results = context.get("concurrent_results", [])
    assert results, "No concurrent results found"

    for response, error in results:
        assert error is None, f"Request failed with error: {error}"
        assert response is not None, "Response is None"
        assert (
            response.status_code == 200
        ), f"Expected status 200, got {response.status_code}"


@then("all responses should indicate SERVING status")
def check_all_responses_serving(context):
    """Check that all concurrent responses indicate SERVING status."""
    results = context.get("concurrent_results", [])
    assert results, "No concurrent results found"

    for response, error in results:
        assert error is None, f"Request failed with error: {error}"
        assert response is not None, "Response is None"

        try:
            json_data = response.json()
        except json.JSONDecodeError:
            assert False, "Response is not valid JSON"

        assert "status" in json_data, "Response does not contain 'status' field"
        assert (
            json_data["status"] == "SERVING_STATUS_SERVING"
        ), f"Expected SERVING_STATUS_SERVING, got {json_data['status']}"


@then("the response should contain proper Connect-Go headers")
def check_connect_go_headers(context):
    """Check that the response contains proper Connect-Go headers."""
    response = context.get("response")
    assert response is not None, "No response received"

    # Check for Connect-Go specific headers
    headers = response.headers
    content_type = headers.get("Content-Type", "")

    # Connect-Go typically uses application/json for HTTP/JSON
    assert (
        "application/json" in content_type
    ), f"Expected JSON content type for Connect-Go, got: {content_type}"

File: test/e2e/steps/connect_steps.go
package steps

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// Connect-Go specific When Steps

// WhenIMakeHealthCheckRequestUsingConnectGo makes a health check request using Connect-Go protocol
func WhenIMakeHealthCheckRequestUsingConnectGo(ctx *TestContext) {
	url := fmt.Sprintf("%s/health.v1.HealthService/Check", ctx.ServerURL)

	req, err := http.NewRequest("POST", url, strings.NewReader("{}"))
	if err != nil {
		ctx.Error = err
		return
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := ctx.HTTPClient.Do(req)
	ctx.Response = resp
	ctx.Error = err

	if resp != nil {
		body, readErr := io.ReadAll(resp.Body)
		if readErr != nil {
			ctx.Error = readErr
		} else {
			ctx.ResponseBody = body
		}
		resp.Body.Close() // nolint:errcheck
	}
}

// WhenIMakeHealthCheckRequestUsingHTTPJSON makes a health check request using HTTP/JSON
func WhenIMakeHealthCheckRequestUsingHTTPJSON(ctx *TestContext) {
	// This is essentially the same as Connect-Go for HTTP/JSON
	WhenIMakeHealthCheckRequestUsingConnectGo(ctx)
}

// WhenIMakeHealthCheckRequestWithInvalidContentType makes a request with invalid content type
func WhenIMakeHealthCheckRequestWithInvalidContentType(ctx *TestContext) {
	url := fmt.Sprintf("%s/health.v1.HealthService/Check", ctx.ServerURL)

	req, err := http.NewRequest("POST", url, strings.NewReader("{}"))
	if err != nil {
		ctx.Error = err
		return
	}

	req.Header.Set("Content-Type", "text/plain")

	resp, err := ctx.HTTPClient.Do(req)
	ctx.Response = resp
	ctx.Error = err

	if resp != nil {
		body, readErr := io.ReadAll(resp.Body)
		if readErr != nil {
			ctx.Error = readErr
		} else {
			ctx.ResponseBody = body
		}
		resp.Body.Close() // nolint:errcheck
	}
}

// WhenIMakeRequestToNonExistentEndpoint makes a request to a non-existent endpoint
func WhenIMakeRequestToNonExistentEndpoint(ctx *TestContext) {
	url := fmt.Sprintf("%s/health.v1.HealthService/NonExistentMethod", ctx.ServerURL)

	req, err := http.NewRequest("POST", url, strings.NewReader("{}"))
	if err != nil {
		ctx.Error = err
		return
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := ctx.HTTPClient.Do(req)
	ctx.Response = resp
	ctx.Error = err

	if resp != nil {
		body, readErr := io.ReadAll(resp.Body)
		if readErr != nil {
			ctx.Error = readErr
		} else {
			ctx.ResponseBody = body
		}
		resp.Body.Close() // nolint:errcheck
	}
}

// WhenIMakeConcurrentHealthCheckRequests makes multiple concurrent health check requests
func WhenIMakeConcurrentHealthCheckRequests(ctx *TestContext, numRequests int) {
	url := fmt.Sprintf("%s/health.v1.HealthService/Check", ctx.ServerURL)

	var wg sync.WaitGroup
	results := make([]ConcurrentResult, numRequests)

	for i := 0; i < numRequests; i++ {
		wg.Add(1)
		go func(index int) {
			defer wg.Done()

			client := &http.Client{Timeout: 5 * time.Second}

			req, err := http.NewRequest("POST", url, strings.NewReader("{}"))
			if err != nil {
				results[index] = ConcurrentResult{Error: err}
				return
			}

			req.Header.Set("Content-Type", "application/json")

			resp, err := client.Do(req)

			result := ConcurrentResult{
				Response: resp,
				Error:    err,
			}

			if resp != nil {
				body, readErr := io.ReadAll(resp.Body)
				if readErr != nil {
					result.Error = readErr
				} else {
					result.Body = body
				}
				resp.Body.Close() // nolint:errcheck
			}

			results[index] = result
		}(i)
	}

	wg.Wait()
	ctx.ConcurrentResults = results
}

// Connect-Go specific Then Steps - BDD Style

// ThenResponseShouldIndicateServingStatus checks that response indicates SERVING status in BDD style
func ThenResponseShouldIndicateServingStatus(ctx *TestContext) {
	ctx.TheResponseStatusCodeShouldBe(http.StatusOK)
	ctx.TheResponseShouldNotBeEmpty()
	ctx.TheJSONResponseShouldContainField("status", "SERVING_STATUS_SERVING")
}

// ThenJSONResponseShouldContainServingStatusServing checks for SERVING_STATUS_SERVING in JSON in BDD style
func ThenJSONResponseShouldContainServingStatusServing(ctx *TestContext) {
	ctx.TheJSONResponseShouldContainField("status", "SERVING_STATUS_SERVING")
}

// ThenServerShouldRejectRequest checks that server rejected the request in BDD style
func ThenServerShouldRejectRequest(ctx *TestContext) {
	ctx.TheResponseStatusCodeShouldBe(http.StatusUnsupportedMediaType)
}

// ThenResponseShouldContainProperConnectGoHeaders checks for proper Connect-Go headers in BDD style
func ThenResponseShouldContainProperConnectGoHeaders(ctx *TestContext) {
	ctx.TheResponseHeaderShouldContain("Content-Type", "application/json")
}

// Legacy functions for backward compatibility
func ThenResponseShouldIndicateServingStatusLegacy(ctx *TestContext) {
	ctx.AssertStatusCode(http.StatusOK)
	ctx.AssertResponseNotEmpty()

	var jsonData map[string]interface{}
	err := json.Unmarshal(ctx.ResponseBody, &jsonData)
	require.NoError(ctx.T, err, "Response is not valid JSON")

	status, exists := jsonData["status"]
	require.True(ctx.T, exists, "Response does not contain 'status' field")
	assert.Equal(ctx.T, "SERVING_STATUS_SERVING", status,
		"Expected SERVING_STATUS_SERVING, got %v", status)
}

// ThenAllResponsesShouldIndicateServingStatus checks all concurrent responses for SERVING status in BDD style
func ThenAllResponsesShouldIndicateServingStatus(ctx *TestContext) {
	if len(ctx.ConcurrentResults) == 0 {
		ctx.T.Errorf("Expected concurrent results to exist, but none were found")
		return
	}

	failedCount := 0
	for i, result := range ctx.ConcurrentResults {
		if result.Error != nil {
			ctx.T.Errorf("Expected concurrent request %d to succeed, but got error: %v", i+1, result.Error)
			failedCount++
			continue
		}
		if result.Response == nil {
			ctx.T.Errorf("Expected concurrent request %d to have a response, but none was received", i+1)
			failedCount++
			continue
		}
		if result.Response.StatusCode != http.StatusOK {
			ctx.T.Errorf("Expected concurrent request %d to have status 200, but got %d", i+1, result.Response.StatusCode)
			failedCount++
			continue
		}

		var jsonData map[string]interface{}
		err := json.Unmarshal(result.Body, &jsonData)
		if err != nil {
			ctx.T.Errorf("Expected concurrent request %d response to be valid JSON, but got error: %v", i+1, err)
			failedCount++
			continue
		}

		status, exists := jsonData["status"]
		if !exists {
			ctx.T.Errorf("Expected concurrent request %d response to contain 'status' field, but it was missing", i+1)
			failedCount++
			continue
		}
		if status != "SERVING_STATUS_SERVING" {
			ctx.T.Errorf("Expected concurrent request %d to have SERVING_STATUS_SERVING, but got %v", i+1, status)
			failedCount++
		}
	}

	if failedCount > 0 {
		ctx.T.Errorf("Expected all %d concurrent requests to indicate serving status, but %d failed", len(ctx.ConcurrentResults), failedCount)
	}
}

// ThenResponseShouldContainProperConnectGoHeadersLegacy checks for proper Connect-Go headers (legacy)
func ThenResponseShouldContainProperConnectGoHeadersLegacy(ctx *TestContext) {
	require.NotNil(ctx.T, ctx.Response, "No response received")

	contentType := ctx.Response.Header.Get("Content-Type")
	assert.Contains(ctx.T, contentType, "application/json",
		"Expected JSON content type for Connect-Go, got: %s", contentType)
}
